### DISCUSSION TITLE
16 x 16

### DISCUSSION THREAD

**Discussion by Kerie70** [zippity21/Stardew_Valley_Regression_Mod#27] [General]
So I was curious, is the 16x16 pixels a hard cap for the sprites or can it be unlocked and made larger, if even slightly?  I would love to make more diapers but the ideas I have floating around in my head only really work if I have more room to add in smaller details.

**Comment by zippity21** [comment]
Looking at the game's built in sprite sheets, I'd say that while, 16x16 is the expectation, you can make them larger. For the Diaper's HUD, we could probably take advantage of the larger size, but it'd get shrunk down to fit in the inventory, shops, etc.

If you're willing to make the sprites (Probably 32x32), I'd be willing to modify the code to use them, as an experiment.

Since all the sizes would need to be increased, I'll attach a starting point. You can make your edits based on this and repost for the experiment (probably won't be able to get to it until tonight.)
![sprites_w_heart_resized](https://user-images.githubusercontent.com/92954308/139851944-bfb11ea9-c53e-404e-8fa5-c20d83204504.png)

**Reply by zippity21** [reply]
I took some time and attempted to clean up the sprite drawing code. I think its in decent shape now. It no longer assumes 16x16 sprites, but the number is still a constant (but now you only have to change it in 1 instead of 20). I actually set it to 64x64 which should allow for a fair amount of detail. This is not in main yet though, since we need a new spritesheet to match new resolution. I can use an upscaler (like I did for the 32x32 above), but it leaves fuzziness and weird artifacts.

It'd be awesome if you, or someone else would be willing to take on redrawing at the higher resolution.

**Reply by Kerie70** [reply]
To redraw every single one?  That make take some time, my pixel art skills are 0.  Took me 2+ hours just to make that one diaper set and that was a recolor copy of the blue space ones.  Making them all larger will take me a fair while assuming I can do it at all.  Since I can't test it client side for myself since I'm still locked at 16x16 I'm not sure how they'll turn out and look in game.

I am already attempting to make a new one at the 32x32 setting you listed above and it is taking me quite a bit of effort.  If you're setting it to 64x64 I can attempt to enlarge them but make no promises. 

Edit:  I have found a way to enlarge them and keep them looking somewhat "sharp" but yeah, I can't test anything larger than 16x16 on my client.

**Comment by Kerie70** [comment]
I do have a question about the upscaling, have you already tested the 32x32 in game and seen how big it'll make the icon on the right side of the screen?   Is 64x64 going to blow it up so large it consumes too much of the screen or will it scale it down some but still keep to detail somehow?

**Reply by zippity21** [reply]
You are correct in that despite being larger images, they still have the same size on-screen. They retain the additional detail because our computer screens are much higher pixel density than the assets. 

Not actual numbers, but to illustrate:
Let's say you want something to show up as 1 inch x 1 inch on the screen
Your monitor may have something like 600 pixels per inch, so it can hold accurate pixel information for an asset that is 600x600 pixels in that square inch.
It just so happens that the game uses less dense images, say 16x16.
So to make the 16x16 asset take up the desired inch on the monitor, each pIxel of the asset becomes about a 38x38 square.
So for a 32x32 asset, each pIxel becomes about a 19x19 square, and so on.
So you don't actually lose any detail until the asset's dimensions exceed your pixel density.

Though it's a point well taken that to redraw them from scratch is a big task. And now that I'm thinking about it, 64x64 may be too big anyway. Going to 32x32 still allows for 4x the detail, but only requires drawing up to 1024px instead of 4096px (though let me know what you think). And, yeah, we can start with the upscaled sheet and just clean it up a little instead of completely redrawing, at least for now.

Being able to test it is a trickier issue. Technically,you could download the source from the new branch and build it yourself, but it does require a little setup. Or I could make a beta release, but I'm not sure if I can do that per branch (maybe I can?). Or, I think GitHub has a way to recompile right from the browser (which would make it pretty easy to test code on a branch), but I have to look into it.

**Reply by zippity21** [reply]
With 64x64, you have enough detail to write 8px text legibly, so it should allow for lots of detail. Don't worry, this was just for testing purposes.
![Capture](https://user-images.githubusercontent.com/92954308/140095002-96e8c920-69e0-4d26-8725-cb614bc4b675.PNG)
 Attaching screenshot.

**Reply by Kerie70** [reply]
That's 8px?  Oh my, I was still using 1-2px for the art on the diapers themselves.

**Reply by zippity21** [reply]
Yeah, since its an 8pt font, the height of a letter is 8px, which makes it quite legible (almost too legible for the game style). So you can get a lot of fine detail since you can draw individual pixels.

**Reply by Kerie70** [reply]
Sorry for the necro post, know it's been awhile, life sucks.  I wanted to come back and work on some new stuff for the mod, diapers, clothes, furniture, etc and was wondering if this increase on the pixels was ever fully implemented or if you can DM me a guide to how you did it for your client?

**Reply by oniryuuko** [reply]
Let me try to get just the simple guide, based solely off of what I'm interpreting from the previous post:
The base game images are 19x19; 16x16 of this is the image itself, the rest of it is some sort of margin.
To upscale, just keep the ratio. This is part of the base game and requires minimal additional coding, probably intended to make 'HD' texture packs easy; I just don't think anyone ever wound up interested in such a work heavy task.
It must always be incremented by a factor of two on the actual image pixels.
So:
Base: 16x16 within a 19x19 space being doubled becomes 32x32 within a 38x38 space, as you're simply doubling all numbers involved.
Four times: 64x64 within a 76x76 space.
Etc, etc.
Just make sure to upscale linearly, such that whatever the margins are, they don't change.

Edit: And keep in mind the target audience. If the size becomes too large, you may begin to lose details to pixel density on smaller screens.
Edit 2: Derp. Missed a comment. The actual code needs a minor change to accept higher resolution textures. Been months since I've done this stuff. >.>

**Reply by zippity21** [reply]
It's been a hot minute for me as well 😅

If I recall correctly, I enabled 64x64 textures with #33 , so it should be available in any releases after that landed.

However, that only allows you to modify existing sprites. You can add more underwear (without recompiling) by extending the sprite image and adding an entry to en.json.

I don't currently have a method to implement other objects.

**Reply by Kerie70** [reply]
I'm missing something, resizing up to 64x64 is not giving me the detail you were showing in your previous screenshot.

**Reply by zippity21** [reply]
Hmmm... I just checked the asset sheet in the current release is 256x768, which corresponds with 64x64 assets.

Keep in mind that I did not keep the text on the diaper for the release, it was just an example.

Unfortunately, my computer is out of commission right now (waiting for a part). Hopefully you can figure out what's going on. If not I should be able to get on it within a week.
